tinygrad/ops.py

```diff
@@ -144,3 +145,31 @@ def truncate_fp16(x):
   dtypes.int32: lambda x: ctypes.c_int32(x).value, dtypes.int64: lambda x: ctypes.c_int64(x).value,}
 
 def exec_alu(op:Op, dtype:DType, operands): return truncate.get(dtype, lambda x: x)(python_alu[op](*operands))
+
+# the living definition of LazyOps
+def verify_lazyop(*ast:LazyOp):
+  children: DefaultDict[LazyOp, Dict[LazyOp, None]] = defaultdict(dict)
+  in_degree: DefaultDict[LazyOp, int] = defaultdict(int)
+  for i, out in enumerate(ast):
+    assert out.arg.idx == i, f"unexpected output buffer idx {out.arg.idx} != {i}"
+    assert out.op is BufferOps.STORE, f"kernels must have stores as the output, got {out.op}"
+    assert out.arg.st.size == ast[-1].arg.st.size, f"outputs must have the same size, got {out.arg.st.size}"
+    for op in out.lazyops:
+      for s in op.src:
+        children[s][op] = None
+        in_degree[op] += 1
+  q = deque((op, op.arg.st) for out in ast for op in out.lazyops if in_degree[op] == 0)
+  sts: Dict[LazyOp, ShapeTracker] = {}
+  while q:
```

Comment:
It's probably simpler to write this as a recursive function with functools.lru_cache, maybe simpler to reason about?